pkg/storage/redis/redis.go

```diff
@@ -208,6 +212,75 @@ func (db *DB) ClearDeleted(_ context.Context, cutoff time.Time) {
 	}
 }
 
+func doNotify(ctx context.Context, psc *redis.PubSubConn, ch chan struct{}) error {
+	switch v := psc.ReceiveWithTimeout(time.Second).(type) {
+	case redis.Message:
+		log.Debug().Str("action", string(v.Data)).Msg("Got redis message")
+		if string(v.Data) != watchAction {
+			return nil
+		}
+		select {
+		case <-ctx.Done():
+			log.Warn().Err(ctx.Err()).Msg("unable to notify channel")
+			return ctx.Err()
+		case ch <- struct{}{}:
+		}
+	case error:
+		log.Debug().Err(v).Msg("redis subscribe error")
+		return v
+	}
+	return nil
+}
+
+// Watch returns a channel to the caller, when there is a change to the version set,
+// sending message to the channel to notify the caller.
+func (db *DB) Watch(ctx context.Context) chan struct{} {
+	ch := make(chan struct{})
+	go func() {
+		c := db.pool.Get()
+		defer func() {
+			close(ch)
+		}()
+
+		// Setup notifications, we only care about changes to db.version_set.
+		if _, err := c.Do("CONFIG", "SET", "notify-keyspace-events", "Kz"); err != nil {
+			log.Error().Err(err).Msg("failed to setup redis notification")
+			c.Close()
+			return
+		}
+
+		c.Close()
+		eb := backoff.NewExponentialBackOff()
+
+	top:
+		psConn := db.pool.Get()
+		defer psConn.Close()
+		psc := redis.PubSubConn{Conn: psConn}
+		if err := psc.PSubscribe("__keyspace*__:" + db.versionSet); err != nil {
+			log.Error().Err(err).Msg("failed to subscribe to version set channel")
+			return
+		}
+
+		for {
+			err, ok := doNotify(ctx, &psc, ch).(net.Error)
+			if !ok && err != nil {
+				log.Error().Err(ctx.Err()).Msg("failed to notify channel")
+				return
+			}
+			if ok && err.Timeout() {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(eb.NextBackOff()):
+				}
+				goto top
```

Comment:
Can we rewrite this without the goto?  I have a very hard time mentally walking through the control flow here.